sql/pgwire: TestAuthenticationAndHBARules failed

[(sql/pgwire).TestAuthenticationAndHBARules failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2501021&tab=buildLog) on [release-20.2@bc7edb032d1389063a08b09c9e55c385ae5f15a0](https://github.com/cockroachdb/cockroach/commits/bc7edb032d1389063a08b09c9e55c385ae5f15a0):

```
                    auth_test.go:260: attempting to scan logs...
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760956494000 Goroutine:38308 File:sql/pgwire/conn.go Line:229 Message:15 session terminated; duration: 184.076787ms Tags:n1,client=127.0.0.1:42316,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760982635000 Goroutine:38308 File:sql/pgwire/server.go Line:505 Message:16 disconnected; duration: 221.38036ms Tags:n1,client=127.0.0.1:42316,hostssl Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760991329000 Goroutine:38200 File:sql/pgwire/server.go Line:497 Message:17 received connection Tags:n1,client=127.0.0.1:42320 Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761018418000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761029651000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761207481000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:260: attempting to scan logs...
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760956494000 Goroutine:38308 File:sql/pgwire/conn.go Line:229 Message:15 session terminated; duration: 184.076787ms Tags:n1,client=127.0.0.1:42316,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760982635000 Goroutine:38308 File:sql/pgwire/server.go Line:505 Message:16 disconnected; duration: 221.38036ms Tags:n1,client=127.0.0.1:42316,hostssl Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760991329000 Goroutine:38200 File:sql/pgwire/server.go Line:497 Message:17 received connection Tags:n1,client=127.0.0.1:42320 Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761018418000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761029651000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761207481000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:260: attempting to scan logs...
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760982635000 Goroutine:38308 File:sql/pgwire/server.go Line:505 Message:16 disconnected; duration: 221.38036ms Tags:n1,client=127.0.0.1:42316,hostssl Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760991329000 Goroutine:38200 File:sql/pgwire/server.go Line:497 Message:17 received connection Tags:n1,client=127.0.0.1:42320 Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761018418000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761029651000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761207481000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761222808000 Goroutine:38200 File:sql/pgwire/conn.go Line:229 Message:21 session terminated; duration: 204.824147ms Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:260: attempting to scan logs...
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760982635000 Goroutine:38308 File:sql/pgwire/server.go Line:505 Message:16 disconnected; duration: 221.38036ms Tags:n1,client=127.0.0.1:42316,hostssl Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760991329000 Goroutine:38200 File:sql/pgwire/server.go Line:497 Message:17 received connection Tags:n1,client=127.0.0.1:42320 Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761018418000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761029651000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761207481000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761222808000 Goroutine:38200 File:sql/pgwire/conn.go Line:229 Message:21 session terminated; duration: 204.824147ms Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:260: attempting to scan logs...
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322760991329000 Goroutine:38200 File:sql/pgwire/server.go Line:497 Message:17 received connection Tags:n1,client=127.0.0.1:42320 Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761018418000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761029651000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761207481000 Goroutine:38201 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761222808000 Goroutine:38200 File:sql/pgwire/conn.go Line:229 Message:21 session terminated; duration: 204.824147ms Tags:n1,client=127.0.0.1:42320,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:282: found log entry: {Severity:INFO Time:1607322761230338000 Goroutine:38200 File:sql/pgwire/server.go Line:505 Message:22 disconnected; duration: 239 Tags:n1,client=127.0.0.1:42320,hostssl Counter:0 Redactable:false}
                    datadriven.go:243: output didn't match expected:
                        @@ -1,7 +1,7 @@
                         I: [n1,client=XXX] 17 received connection
                         I: [n1,client=XXX,hostssl,user=root] 18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule
                         I: [n1,client=XXX,hostssl,user=root] 19 no client certificate, proceeding with password authentication
                         I: [n1,client=XXX,hostssl,user=root] 20 authentication failed: password authentication failed for user root
                         I: [n1,client=XXX,hostssl,user=root] 21 session terminated; duration: XXX
                        -I: [n1,client=XXX,hostssl] 22 disconnected; duration: XXX
                        +I: [n1,client=XXX,hostssl] 22 disconnected; duration: 239
                         
=== RUN   TestAuthenticationAndHBARules/insecure=false/conn_log/conn_tls
            --- FAIL: TestAuthenticationAndHBARules/insecure=false/conn_log/conn_tls (0.53s)
=== RUN   TestAuthenticationAndHBARules/insecure=false/conn_log
        --- FAIL: TestAuthenticationAndHBARules/insecure=false/conn_log (1.35s)
            test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestAuthenticationAndHBARules_insecure=false_conn_log253567018

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestAuthenticationAndHBARules PKG=./pkg/sql/pgwire TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAuthenticationAndHBARules.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/pgwire).TestAuthenticationAndHBARules failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2914079&tab=buildLog) on [release-20.2@0f05d39f4eca3b1f7eee8a7b391fcbcdac58a9d4](https://github.com/cockroachdb/cockroach/commits/0f05d39f4eca3b1f7eee8a7b391fcbcdac58a9d4):

```
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756416427000 Goroutine:37347 File:sql/pgwire/auth.go Line:335 Message:12 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756416588000 Goroutine:37347 File:sql/pgwire/auth.go Line:335 Message:13 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756681782000 Goroutine:37347 File:sql/pgwire/auth.go Line:335 Message:14 authentication succeeded Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710717000 Goroutine:37334 File:sql/pgwire/conn.go Line:233 Message:15 session terminated; duration: 294.367569ms Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710884000 Goroutine:37334 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 308.233614ms Tags:n1,client=127.0.0.1:37198,hostssl Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710717000 Goroutine:37334 File:sql/pgwire/conn.go Line:233 Message:15 session terminated; duration: 294.367569ms Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710884000 Goroutine:37334 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 308.233614ms Tags:n1,client=127.0.0.1:37198,hostssl Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756713847000 Goroutine:37350 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:37210 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720521000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720617000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756983848000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710717000 Goroutine:37334 File:sql/pgwire/conn.go Line:233 Message:15 session terminated; duration: 294.367569ms Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710884000 Goroutine:37334 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 308.233614ms Tags:n1,client=127.0.0.1:37198,hostssl Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756713847000 Goroutine:37350 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:37210 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720521000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720617000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756983848000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710717000 Goroutine:37334 File:sql/pgwire/conn.go Line:233 Message:15 session terminated; duration: 294.367569ms Tags:n1,client=127.0.0.1:37198,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756710884000 Goroutine:37334 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 308.233614ms Tags:n1,client=127.0.0.1:37198,hostssl Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756713847000 Goroutine:37350 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:37210 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720521000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720617000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756983848000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756713847000 Goroutine:37350 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:37210 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720521000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756720617000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756983848000 Goroutine:37233 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756991123000 Goroutine:37350 File:sql/pgwire/conn.go Line:233 Message:21 session terminated; duration: 270.648024ms Tags:n1,client=127.0.0.1:37210,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1619128756991347000 Goroutine:37350 File:sql/pgwire/server.go Line:548 Message:22 disconnected; duration: 27 Tags:n1,client=127.0.0.1:37210,hostssl Counter:0 Redactable:false}
                    datadriven.go:243: output didn't match expected:
                        @@ -1,7 +1,7 @@
                         I: [n1,client=XXX] 17 received connection
                         I: [n1,client=XXX,hostssl,user=root] 18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule
                         I: [n1,client=XXX,hostssl,user=root] 19 no client certificate, proceeding with password authentication
                         I: [n1,client=XXX,hostssl,user=root] 20 authentication failed: password authentication failed for user root
                         I: [n1,client=XXX,hostssl,user=root] 21 session terminated; duration: XXX
                        -I: [n1,client=XXX,hostssl] 22 disconnected; duration: XXX
                        +I: [n1,client=XXX,hostssl] 22 disconnected; duration: 27
                         
=== RUN   TestAuthenticationAndHBARules/insecure=false/conn_log/conn_tls
            --- FAIL: TestAuthenticationAndHBARules/insecure=false/conn_log/conn_tls (0.62s)
=== RUN   TestAuthenticationAndHBARules/insecure=false/conn_log
        --- FAIL: TestAuthenticationAndHBARules/insecure=false/conn_log (2.02s)
            test_log_scope.go:158: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestAuthenticationAndHBARules_insecure=false_conn_log323153807
=== RUN   TestAuthenticationAndHBARules/insecure=false
    --- FAIL: TestAuthenticationAndHBARules/insecure=false (14.37s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestAuthenticationAndHBARules PKG=./pkg/sql/pgwire TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAuthenticationAndHBARules.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/pgwire).TestAuthenticationAndHBARules failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3060669&tab=buildLog) on [release-20.2@4b7eb8b8df8f328ca0077a9bec7d2c8f502aaa63](https://github.com/cockroachdb/cockroach/commits/4b7eb8b8df8f328ca0077a9bec7d2c8f502aaa63):

```
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849358665000 Goroutine:38646 File:sql/pgwire/auth.go Line:335 Message:12 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849358761000 Goroutine:38646 File:sql/pgwire/auth.go Line:335 Message:13 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849542769000 Goroutine:38646 File:sql/pgwire/auth.go Line:335 Message:14 authentication succeeded Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849352437000 Goroutine:38614 File:sql/pgwire/server.go Line:540 Message:11 received connection Tags:n1,client=127.0.0.1:49942 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849358665000 Goroutine:38646 File:sql/pgwire/auth.go Line:335 Message:12 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849358761000 Goroutine:38646 File:sql/pgwire/auth.go Line:335 Message:13 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849542769000 Goroutine:38646 File:sql/pgwire/auth.go Line:335 Message:14 authentication succeeded Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849564942000 Goroutine:38614 File:sql/pgwire/conn.go Line:234 Message:15 session terminated; duration: 206.32331ms Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849565077000 Goroutine:38614 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 212.639689ms Tags:n1,client=127.0.0.1:49942,hostssl Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849564942000 Goroutine:38614 File:sql/pgwire/conn.go Line:234 Message:15 session terminated; duration: 206.32331ms Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849565077000 Goroutine:38614 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 212.639689ms Tags:n1,client=127.0.0.1:49942,hostssl Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849570711000 Goroutine:37927 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:49956 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575545000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575640000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849758304000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849564942000 Goroutine:38614 File:sql/pgwire/conn.go Line:234 Message:15 session terminated; duration: 206.32331ms Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849565077000 Goroutine:38614 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 212.639689ms Tags:n1,client=127.0.0.1:49942,hostssl Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849570711000 Goroutine:37927 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:49956 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575545000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575640000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849758304000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849564942000 Goroutine:38614 File:sql/pgwire/conn.go Line:234 Message:15 session terminated; duration: 206.32331ms Tags:n1,client=127.0.0.1:49942,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849565077000 Goroutine:38614 File:sql/pgwire/server.go Line:548 Message:16 disconnected; duration: 212.639689ms Tags:n1,client=127.0.0.1:49942,hostssl Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849570711000 Goroutine:37927 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:49956 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575545000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575640000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849758304000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:262: attempting to scan logs...
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849570711000 Goroutine:37927 File:sql/pgwire/server.go Line:540 Message:17 received connection Tags:n1,client=127.0.0.1:49956 Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575545000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849575640000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:19 no client certificate, proceeding with password authentication Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849758304000 Goroutine:37928 File:sql/pgwire/auth.go Line:335 Message:20 authentication failed: password authentication failed for user root Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849764681000 Goroutine:37927 File:sql/pgwire/conn.go Line:234 Message:21 session terminated; duration: 189.174733ms Tags:n1,client=127.0.0.1:49956,hostssl,user=root Counter:0 Redactable:false}
                    auth_test.go:284: found log entry: {Severity:INFO Time:1623164849764908000 Goroutine:37927 File:sql/pgwire/server.go Line:548 Message:22 disconnected; duration: 19 Tags:n1,client=127.0.0.1:49956,hostssl Counter:0 Redactable:false}
                    datadriven.go:243: output didn't match expected:
                        @@ -1,7 +1,7 @@
                         I: [n1,client=XXX] 17 received connection
                         I: [n1,client=XXX,hostssl,user=root] 18 connection matches HBA rule: host  all root all cert-password # CockroachDB mandatory rule
                         I: [n1,client=XXX,hostssl,user=root] 19 no client certificate, proceeding with password authentication
                         I: [n1,client=XXX,hostssl,user=root] 20 authentication failed: password authentication failed for user root
                         I: [n1,client=XXX,hostssl,user=root] 21 session terminated; duration: XXX
                        -I: [n1,client=XXX,hostssl] 22 disconnected; duration: XXX
                        +I: [n1,client=XXX,hostssl] 22 disconnected; duration: 19
                         
=== RUN   TestAuthenticationAndHBARules/insecure=false/conn_log/conn_tls
            --- FAIL: TestAuthenticationAndHBARules/insecure=false/conn_log/conn_tls (0.43s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestAuthenticationAndHBARules PKG=./pkg/sql/pgwire TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAuthenticationAndHBARules.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/util/log/file.go
+++ b/pkg/util/log/file.go
@@ -95,7 +95,7 @@ type fileSink struct {
 	// mu protects the remaining elements of this structure and is
 	// used to synchronize output to this file sink..
 	mu struct {
-		syncutil.Mutex
+		syncutil.RWMutex
 
 		// directory prefix where to store this logger's files. This is
 		// under "mu" because the test Scope can overwrite this

--- a/pkg/util/log/file_api.go
+++ b/pkg/util/log/file_api.go
@@ -21,6 +21,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/util/log/logpb"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/errors/oserror"
 )
@@ -210,15 +211,18 @@ func GetLogReader(filename string) (io.ReadCloser, error) {
 	if fs == nil || !fs.enabled.Get() {
 		return nil, errors.Newf("no log directory found for %s", filename)
 	}
-	fs.mu.Lock()
-	dir := fs.mu.logDir
-	fs.mu.Unlock()
+	dir := func() string {
+		fs.mu.RLock()
+		defer fs.mu.RUnlock()
+		return fs.mu.logDir
+	}()
 	if dir == "" {
 		// This error should never happen: .enabled should be unset in
 		// that case.
 		return nil, errors.Newf("no log directory found for %s", filename)
 	}
 
+	baseFileName := filename
 	filename = filepath.Join(dir, filename)
 
 	info, err := os.Lstat(filename)
@@ -236,7 +240,23 @@ func GetLogReader(filename string) (io.ReadCloser, error) {
 		return nil, errors.Errorf("not a regular file")
 	}
 
-	return os.Open(filename)
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	sb, ok := fs.mu.file.(*syncBuffer)
+	if ok && baseFileName == filepath.Base(sb.file.Name()) {
+		// If the file being read is also the file being written to, then we
+		// want mutual exclusion between the reader and the flusher.
+		lr := &lockedReader{}
+		lr.mu.RWMutex = &fs.mu.RWMutex
+		lr.mu.wrappedFile = file
+		return lr, nil
+	}
+	return file, nil
 }
 
 // sortablelogpb.FileInfoSlice is required so we can sort logpb.FileInfos.
@@ -346,6 +366,40 @@ func (a sortableEntries) Less(i, j int) bool {
 	return a[i].Time > a[j].Time
 }
 
+var _ io.ReadCloser = (*lockedReader)(nil)
+
+// lockedReader locks accesses to a wrapped io.ReadCloser,
+// using a RWMutex shared with another component.
+// We use this when reading log files (using the GetLogReader API)
+// that are concurrently being written to by the log flusher,
+// to ensure that read operations cannot observe partial flushes.
+type lockedReader struct {
+	mu struct {
+		// We use a mutex by reference, so that we can point this
+		// lockedReader to the same mutex as used by the corresponding
+		// fileSink.
+		// This mutex is only defined if the file being read from
+		// can also be written to concurrently.
+		*syncutil.RWMutex
+
+		wrappedFile io.ReadCloser
+	}
+}
+
+func (r *lockedReader) Read(b []byte) (int, error) {
+	if r.mu.RWMutex != nil {
+		r.mu.RLock()
+		defer r.mu.RUnlock()
+	}
+	return r.mu.wrappedFile.Read(b)
+}
+
+func (r *lockedReader) Close() error {
+	// We do not need to hold the mutex to call Close() since there is
+	// no flushing needed on read-only file access during Close() calls.
+	return r.mu.wrappedFile.Close()
+}
+
 // readAllEntriesFromFile reads in all log entries from a given file that are
 // between the 'startTimestamp' and 'endTimestamp' and match the 'pattern' if it
 // exists. It returns the entries in the reverse chronological order. It also
@@ -366,6 +420,7 @@ func readAllEntriesFromFile(
 		return nil, false, err
 	}
 	defer reader.Close()
+
 	entries := []logpb.Entry{}
 	decoder, err := NewEntryDecoderWithFormat(reader, editMode, format)
 	if err != nil {

